crates/jstz_core/src/kv/transaction.rs

```diff
@@ -43,258 +40,353 @@ use std::rc::Rc;
+    }
+
+    pub fn contains_key(&self, key: &Key) -> bool {
+        self.insert_edits.contains_key(key) && !self.remove_edits.contains(key)
     }
 }
 
-impl Default for Transaction {
-    fn default() -> Self {
-        Self::new()
+impl LookupMap {
+    fn update(&mut self, key: Key, idx: usize) {
+        let key_history = self.entry(key).or_default();
+
+        match key_history.last() {
+            Some(&last_idx) if last_idx == idx => {
+                // The key was already looked up in the current context
+            }
+            _ => {
+                key_history.push(idx);
+            }
+        }
+    }
+
+    fn rollback(&mut self, key: &Key) -> Result<()> {
+        let is_history_empty = {
+            let history = self.get_mut(key).ok_or(KvError::ExpectedLookupMapEntry)?;
+
+            history.pop();
+            history.is_empty()
+        };
+
+        if is_history_empty {
+            self.remove(key);
+        }
+
+        Ok(())
     }
 }
 
 impl Transaction {
-    pub fn new() -> Self {
-        Self {
-            parent: None,
-            remove_set: BTreeSet::new(),
-            snapshot: BTreeMap::new(),
+    fn csnapshot_idx(&self) -> usize {
+        self.stack.len().saturating_sub(1)
+    }
+
+    fn update_lookup_map(&mut self, key: Key) {
+        self.lookup_map.update(key, self.csnapshot_idx())
+    }
+
+    /// Return the current snapshot
+    fn csnapshot(&mut self) -> Result<&mut Snapshot> {
+        Ok(self
+            .stack
+            .last_mut()
+            .ok_or(KvError::TransactionStackEmpty)?)
+    }
+
+    /// Insert a key-value pair into the current snapshot (as a 'insert' edit)
+    fn csnapshot_insert(&mut self, key: Key, value: SnapshotValue) -> Result<()> {
+        self.update_lookup_map(key.clone());
+        self.csnapshot()?.insert(key, value);
+        Ok(())
+    }
+
+    /// Lookup a key in the current snapshot
+    fn csnapshot_lookup(&mut self, key: &Key) -> Result<Option<&SnapshotValue>> {
+        Ok(self.csnapshot()?.lookup(key))
+    }
+
+    /// Lookup a key in the current snapshot
+    fn csnapshot_lookup_mut(&mut self, key: &Key) -> Result<Option<&mut SnapshotValue>> {
+        Ok(self.csnapshot()?.lookup_mut(key))
+    }
+
+    /// Remove a key from the current snapshot (as a 'remove' edit)
+    fn csnapshot_remove(&mut self, key: Key) -> Result<()> {
+        self.update_lookup_map(key.clone());
+        self.csnapshot()?.remove(key);
+        Ok(())
+    }
+
+    fn lookup<V>(&mut self, rt: &impl Runtime, key: Key) -> Result<Option<&SnapshotValue>>
+    where
+        V: Value + DeserializeOwned,
+    {
+        if let Some(&snapshot_idx) =
+            self.lookup_map.get(&key).and_then(|history| history.last())
+        {
+            let snapshot = &self.stack[snapshot_idx];
+
+            return Ok(snapshot.lookup(&key));
+        }
+
+        if let Some(value) = Storage::get::<V>(rt, &key)? {
+            // TODO: This clone is probably not necessary
```

Comment:
please either remove the TODO or clone

Reply:
TODO will be resolved (possibly) in a future PR -- I'm not convinced it can easily be fixed at the moment